File: web/src/lib/queues.ts
/**
 * Queue ids and labels, the single source for queue names in the UI.
 *
 * Riot adds queue ids faster than anyone maps them (queue 710 arrived in August 2026 and is in
 * no static list), so every label falls back to the game mode before it falls back to the raw
 * id: an unmapped Summoner's Rift queue reads "Summoner's Rift", never "Queue 710".
 */
import type { QueueType } from "@/api/types";

export const QUEUE_SOLO = 420;
export const QUEUE_FLEX = 440;
export const RANKED_QUEUES: readonly number[] = [QUEUE_SOLO, QUEUE_FLEX];

export const QUEUE_LABELS: Readonly<Record<number, string>> = {
  0: "Custom",
  400: "Normal Draft",
  420: "Ranked Solo/Duo",
  430: "Normal Blind",
  440: "Ranked Flex",
  450: "ARAM",
  480: "Swiftplay",
  490: "Quickplay",
  700: "Clash",
  720: "ARAM Clash",
  830: "Co-op vs AI",
  840: "Co-op vs AI",
  850: "Co-op vs AI",
  900: "ARURF",
  1700: "Arena",
  1710: "Arena",
  1900: "URF",
};

/** Compact names for dense rows ("Solo/Duo"). */
const SHORT_LABELS: Readonly<Record<number, string>> = {
  400: "Draft",
  420: "Solo/Duo",
  430: "Blind",
  440: "Flex",
  450: "ARAM",
  480: "Swiftplay",
  490: "Quickplay",
  700: "Clash",
  900: "ARURF",
  1700: "Arena",
  1710: "Arena",
  1900: "URF",
};

/** Names for match rows, where "Ranked Solo" reads better than "Ranked Solo/Duo". */
const ROW_LABELS: Readonly<Record<number, string>> = {
  ...QUEUE_LABELS,
  420: "Ranked Solo",
};

/** `game_mode` from match-v5 -> what a player calls it. */
const GAME_MODE_LABELS: Readonly<Record<string, string>> = {
  CLASSIC: "Summoner's Rift",
  ARAM: "ARAM",
  CHERRY: "Arena",
  URF: "URF",
  ARURF: "ARURF",
  ONEFORALL: "One for All",
  NEXUSBLITZ: "Nexus Blitz",
  ULTBOOK: "Ultimate Spellbook",
  STRAWBERRY: "Swarm",
  BRAWL: "Brawl",
  TUTORIAL: "Tutorial",
  PRACTICETOOL: "Practice Tool",
};

/** Compact game-mode names, for the same narrow columns as SHORT_LABELS. */
const GAME_MODE_SHORT: Readonly<Record<string, string>> = {
  CLASSIC: "Rift",
  NEXUSBLITZ: "Blitz",
  ULTBOOK: "Spellbook",
  PRACTICETOOL: "Practice",
};

/** Name of a game mode ("CLASSIC" -> "Summoner's Rift"); null when the mode is unknown too. */
function gameModeLabel(gameMode: string | null | undefined, short = false): string | null {
  const mode = gameMode?.trim().toUpperCase();
  if (!mode) return null;
  return (short ? GAME_MODE_SHORT[mode] : undefined) ?? GAME_MODE_LABELS[mode] ?? null;
}

/** A server label the backend made up for an id it doesn't know ("Queue 710"). */
function isPlaceholderLabel(label: string | null | undefined): boolean {
  return !label || /^queue \d+$/i.test(label.trim());
}

function fallbackLabel(queueId: number, gameMode: string | null | undefined, short = false): string {
  return gameModeLabel(gameMode, short) ?? `Queue ${queueId}`;
}

/**
 * Full queue name. Prefers the API's own label (`MatchSummary.queue_label`, which follows
 * Riot's list) and falls back to the local table, then the game mode.
 */
export function queueLabel(queueId: number, gameMode?: string | null, serverLabel?: string | null): string {
  if (serverLabel && !isPlaceholderLabel(serverLabel)) return serverLabel;
  return QUEUE_LABELS[queueId] ?? fallbackLabel(queueId, gameMode);
}

/** Queue name for a dense match row ("Ranked Solo"). Used by components/match/matchUtils. */
export function queueRowLabel(queueId: number, gameMode?: string | null, serverLabel?: string | null): string {
  return ROW_LABELS[queueId] ?? queueLabel(queueId, gameMode, serverLabel);
}

/** Shortest queue name, for chips and narrow columns ("Solo/Duo"). */
export function queueShortLabel(queueId: number, gameMode?: string | null): string {
  return SHORT_LABELS[queueId] ?? QUEUE_LABELS[queueId] ?? fallbackLabel(queueId, gameMode, true);
}

export function isRankedQueue(queueId: number): boolean {
  return RANKED_QUEUES.includes(queueId);
}

export const QUEUE_TYPE_LABELS: Readonly<Record<QueueType, string>> = {
  RANKED_SOLO_5x5: "Ranked Solo/Duo",
  RANKED_FLEX_SR: "Ranked Flex",
};

File: api/src/hextrack/queues.py
"""Queue id -> label map (https://static.developer.riotgames.com/docs/lol/queues.json)."""

from __future__ import annotations

from typing import Final

RANKED_SOLO: Final = 420
RANKED_FLEX: Final = 440
RANKED_QUEUES: Final[frozenset[int]] = frozenset({RANKED_SOLO, RANKED_FLEX})

QUEUE_LABELS: Final[dict[int, str]] = {
    0: "Custom",
    400: "Normal Draft",
    420: "Ranked Solo/Duo",
    430: "Normal Blind",
    440: "Ranked Flex",
    450: "ARAM",
    480: "Swiftplay",
    490: "Quickplay",
    700: "Clash",
    720: "ARAM Clash",
    830: "Co-op vs AI Intro",
    840: "Co-op vs AI Beginner",
    850: "Co-op vs AI Intermediate",
    870: "Co-op vs AI Intro",
    880: "Co-op vs AI Beginner",
    890: "Co-op vs AI Intermediate",
    900: "ARURF",
    1020: "One for All",
    1300: "Nexus Blitz",
    1400: "Ultimate Spellbook",
    1700: "Arena",
    1710: "Arena",
    1810: "Swarm",
    1820: "Swarm",
    1830: "Swarm",
    1840: "Swarm",
    1900: "URF",
    2300: "Brawl",
    2400: "ARAM: Mayhem",
    3100: "Custom",
}


def queue_label(queue_id: int) -> str:
    """Label for a queue id; unknown ids render as "Queue <id>"."""
    return QUEUE_LABELS.get(queue_id, f"Queue {queue_id}")


def is_ranked(queue_id: int) -> bool:
    return queue_id in RANKED_QUEUES


def queue_labels_json() -> dict[str, str]:
    """``{"420": "Ranked Solo/Duo", ...}`` for the /meta endpoint (JSON keys are strings)."""
    return {str(k): v for k, v in sorted(QUEUE_LABELS.items())}
